src/upload-button.tsx

```diff
@@ -0,0 +1,351 @@
+    const [uploadedFiles, setUploadedFiles] = useState<{[name: string]:
+                                                        {file: File, progress: number, cancel:() => void}}>({});
+
+    const handleClick = () => {
+        if (ref.current) {
+            ref.current.click();
+        }
+    };
+
+    const beforeUnloadHandler = (event: BeforeUnloadEvent) => {
+        event.preventDefault();
+
+        // Included for legacy support, e.g. Chrome/Edge < 119
+        event.returnValue = true;
+    };
+
+    const onUpload = async (event: React.ChangeEvent<HTMLInputElement>) => {
+        cockpit.assert(event.target.files, "not an <input type='file'>?");
+        cockpit.assert(cwdInfo?.entries, "No cwdInfo.entries is undefined");
+        let next_progress = 0;
+        const toUploadFiles = [];
+
+        const resetInput = () => {
+        // Reset input field in the case a download was cancelled and has to be re-uploaded
+        // https://stackoverflow.com/questions/26634616/filereader-upload-same-file-again-not-working
+            event.target.value = "";
+        };
+
+        let resolution;
+        let replaceAll = false;
+        let skipAll = false;
+        for (let i = 0; i < event.target.files.length; i++) {
+            const inputFile = event.target.files[i];
+            const file = cwdInfo?.entries[inputFile.name];
+
+            if (replaceAll)
+                toUploadFiles.push(inputFile);
+            else if (file && skipAll) {
+                continue;
+            } else if (file) {
+                try {
+                    resolution = await Dialogs.show(
+                        <FileConflictDialog
+                          path={path}
+                          file={file}
+                          uploadFile={inputFile}
+                          showApplyAll={event.target.files.length > 1}
+                        />
+                    );
+                } catch (exc) {
+                    resetInput();
+                    return;
+                }
+
+                if (resolution.skip) {
+                    if (resolution.applyToAll)
+                        skipAll = true;
+                    continue;
+                }
+
+                if (resolution.applyToAll && resolution.replace)
+                    replaceAll = true;
+
+                toUploadFiles.push(inputFile);
+            } else {
+                toUploadFiles.push(inputFile);
+            }
+        }
+
+        if (toUploadFiles.length === 0) {
+            resetInput();
+            return;
+        }
+
+        window.addEventListener("beforeunload", beforeUnloadHandler);
+
+        const cancelledUploads = [];
+        await Promise.allSettled(toUploadFiles.map(async (file: File) => {
+            const tmp_path = path.slice();
+            tmp_path.push(file.name);
+            const destination = tmp_path.join('/');
+            const abort = new AbortController();
+
+            setUploadedFiles(oldFiles => {
```

Comment:
The more often I see code like this the more often I want to add a general-purpose reducer based on json-merge to our `hooks.ts`....

Reply:
Hahahaha, yes I agree object state is a bit annoying to work with at times.